Sources/PackageCollectionsSigning/Certificate/Certificate.swift

```diff
@@ -0,0 +1,246 @@
+    }
+
+    private func extractName(_ name: CFString) throws -> CertificateName {
+        guard let dict = SecCertificateCopyValues(self.underlying, [name] as CFArray, nil) as? [CFString: Any],
+            let nameDict = dict[name] as? [CFString: Any],
+            let propValueList = nameDict[kSecPropertyKeyValue] as? [[String: Any]] else {
+            throw CertificateError.nameExtractionFailure
+        }
+
+        let props = propValueList.reduce(into: [String: String]()) { result, item in
+            if let label = item["label"] as? String, let value = item["value"] as? String {
+                result[label] = value
+            }
+        }
+
+        return CertificateName(
+            userID: props["0.9.2342.19200300.100.1.1"], // FIXME: don't hardcode?
+            commonName: props[kSecOIDCommonName as String],
+            organization: props[kSecOIDOrganizationName as String],
+            organizationalUnit: props[kSecOIDOrganizationalUnitName as String]
+        )
+    }
+
+    func publicKey() throws -> PublicKey {
+        guard let key = SecCertificateCopyKey(self.underlying) else {
+            throw CertificateError.keyExtractionFailure
+        }
+
+        var error: Unmanaged<CFError>?
+        guard let data = SecKeyCopyExternalRepresentation(key, &error) as Data? else {
+            throw error.map { $0.takeRetainedValue() as Error } ?? CertificateError.keyExtractionFailure
+        }
+
+        switch try self.keyType(of: key) {
+        case .RSA:
+            return try CoreRSAPublicKey(data: data)
+        case .EC:
+            return try ECPublicKey(data: data)
+        }
+    }
+
+    func keyType() throws -> KeyType {
+        guard let key = SecCertificateCopyKey(self.underlying) else {
+            throw CertificateError.keyExtractionFailure
+        }
+        return try self.keyType(of: key)
+    }
+
+    private func keyType(of key: SecKey) throws -> KeyType {
+        guard let attributes = SecKeyCopyAttributes(key) as? [CFString: Any],
+            let keyType = attributes[kSecAttrKeyType] as? String else {
+            throw CertificateError.indeterminateKeyType
+        }
+
+        if keyType == (kSecAttrKeyTypeRSA as String) {
+            return .RSA
+        } else if keyType == (kSecAttrKeyTypeEC as String) {
+            return .EC
+        } else {
+            throw CertificateError.unsupportedKeyType
+        }
+    }
+}
+
+// MARK: - Certificate implementation using BoringSSL
+
+#else
+final class BoringSSLCertificate {
+    let underlying: UnsafeMutablePointer<X509>
+
+    deinit {
+        CCryptoBoringSSL_X509_free(self.underlying)
+    }
+
+    init(derEncoded data: Data) throws {
+        let bytes = data.copyBytes()
+        let x509 = try bytes.withUnsafeBufferPointer { (ptr: UnsafeBufferPointer<UInt8>) throws -> UnsafeMutablePointer<X509> in
+            var pointer = ptr.baseAddress
+            guard let x509 = CCryptoBoringSSL_d2i_X509(nil, &pointer, bytes.count) else {
```

Comment:
This breaks on Windows due to the type mismatch of `Data.count` being `Int` and `CCryptoBoringSSL_d2i_X509` taking a `CInt`.
```suggestion
            guard let x509 = CCryptoBoringSSL_d2i_X509(nil, &pointer, CInt(bytes.count)) else {
```

Reply:
Addressed in https://github.com/apple/swift-package-manager/pull/3238/commits/07c8a430920b6a3103b8a80179f5767ab69044f4 but use `numericCast` instead of `CInt` since it fails on Linux.